File: frontend/src/lib/api/locations.ts
import { deleteResource, getJson, postJson, putJson } from "./client";

/**
 * A persisted location ("room") as returned by `GET /api/v1/locations`.
 *
 * Server response fields are snake_case (design §1: `created_at`/`updated_at`).
 * No runtime validation is performed; the shape mirrors `LocationResponse`.
 */
export interface Location {
  readonly id: number;
  readonly name: string;
  readonly notes: string | null;
  readonly created_at: string;
  readonly updated_at: string;
}

/** Request body for creating/updating a location (`LocationCreate`/`LocationUpdate`). */
export interface LocationInput {
  readonly name: string;
  readonly notes: string | null;
}

/** `GET /api/v1/locations` - all rooms, ordered by name. Throws `ApiError` on non-2xx. */
export function fetchLocations(): Promise<Location[]> {
  return getJson<Location[]>("/locations");
}

/** `GET /api/v1/locations/{id}` - one room. Throws `ApiError` on non-2xx. */
export function fetchLocation(id: number): Promise<Location> {
  return getJson<Location>(`/locations/${String(id)}`);
}

/** `POST /api/v1/locations` - create a room (201). Throws `ApiError` on non-2xx. */
export function createLocation(input: LocationInput): Promise<Location> {
  return postJson<Location>("/locations", input);
}

/** `PUT /api/v1/locations/{id}` - full-replace update (200). Throws `ApiError` on non-2xx. */
export function updateLocation(
  id: number,
  input: LocationInput,
): Promise<Location> {
  return putJson<Location>(`/locations/${String(id)}`, input);
}

/** `DELETE /api/v1/locations/{id}` - remove a room (204). Throws `ApiError` on non-2xx. */
export function deleteLocation(id: number): Promise<void> {
  return deleteResource(`/locations/${String(id)}`);
}

File: backend/src/viridarium/adapters/inbound/web/locations.py
"""Locations router (ARCH-002: HTTP only, no business logic).

Maps the wire schemas to/from domain types and delegates to the
:class:`~viridarium.application.locations.LocationService`. ``LocationNotFoundError``
is not caught here: it is translated to HTTP 404 by the registered exception handler
in the app factory (ADR-C), keeping status codes out of the router.
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, Response, status

from viridarium.adapters.inbound.web.dependencies import get_location_service
from viridarium.adapters.inbound.web.schemas import (
    LocationCreate,
    LocationResponse,
    LocationUpdate,
)
from viridarium.application.locations import LocationService
from viridarium.domain.location import NewLocation

router = APIRouter(prefix="/locations", tags=["locations"])

ServiceDep = Annotated[LocationService, Depends(get_location_service)]


@router.post(
    "",
    response_model=LocationResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a location",
)
def create_location(body: LocationCreate, service: ServiceDep) -> LocationResponse:
    """Create a new room/location."""
    created = service.create(NewLocation(name=body.name, notes=body.notes))
    return LocationResponse.model_validate(created)


@router.get("", response_model=list[LocationResponse], summary="List locations")
def list_locations(service: ServiceDep) -> list[LocationResponse]:
    """List all rooms/locations ordered by name."""
    return [LocationResponse.model_validate(loc) for loc in service.list()]


@router.get("/{location_id}", response_model=LocationResponse, summary="Get a location")
def get_location(location_id: int, service: ServiceDep) -> LocationResponse:
    """Get one room/location by id."""
    return LocationResponse.model_validate(service.get(location_id))


@router.put(
    "/{location_id}", response_model=LocationResponse, summary="Replace a location"
)
def update_location(
    location_id: int, body: LocationUpdate, service: ServiceDep
) -> LocationResponse:
    """Full-replace a room/location's name and notes (ADR-D)."""
    updated = service.update(location_id, name=body.name, notes=body.notes)
    return LocationResponse.model_validate(updated)


@router.delete(
    "/{location_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a location",
)
def delete_location(location_id: int, service: ServiceDep) -> Response:
    """Delete a room/location by id."""
    service.delete(location_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)
